crypto/bn/bn_rand.c

```diff
@@ -109,6 +109,23 @@ int BN_priv_rand(BIGNUM *rnd, int bits, int top, int bottom)
     return bnrand(PRIVATE, rnd, bits, top, bottom);
 }
 
+/*
+ * Retrieve a string of bits from a private RBG and convert to a BIGNUM.
+ * Repeat this process until the generated BIGNUM is smaller than the 'upper'
+ * bound.
+ */
+int bn_rand_priv_range(BIGNUM *rnd, int bits, BIGNUM *upper)
```

Comment:
Why not use BN_priv_rand_range()? Why the extra bits parameter?

Reply:
The old method of generation just did get a random and check it is range (then internally BN_priv_rand_range() does some very interesting optimizations (involving modulus) when the 2nd and 3rd most top bits are not set). 
The new algorithm is based on grabbing a exact number of bits of random, and then checking if that value is in range.
So for FIPS I am not sure that you could argue that they are the same operation.
A side effect of the new algorithm is that the output value is 1 + rand()... Which breaks some tests which feed the entropy in thru a buffer (which now needs to be subtracted by 1)

Reply:
> The old method of generation just got a random and check it is in range (then internally BN_priv_rand_range() does some very interesting optimizations (involving modulus) when the 2nd and 3rd most top bits are not set).

https://github.com/openssl/openssl/blame/master/crypto/bn/bn_rand.c#L113

`BN_priv_rand_range()` derives the number of bits from the range, so that part is functionally compatible to what is being done here passing the number of bits;
I admit I don't understand why `BN_priv_rand_range()` is checking the 2 and 3 topmost bit of `range` as well and behaving differently depending on that check: it's 18 years old code and I cannot find a motivation looking back at the history of it (seems to be related to Bleichenbacher attacks on biased DSA nonces).

In any case either that behaviour is still required (and the new ad-hoc method is bypassing it) or is ossified legacy code (and we would be better off revising it rather than internally use something different and not accessible to external developers).

Reply:
So the existing one will probably be faster as it grabs 1 more bit of random and then tries some modulus operations (by subtracting the range one or 2 times). I cant tell if this is biased or not. The one I am trying to use doesnt have bias.. 
So the issue it was trying to solve (I think) is this..
Imagine the top bit is set in the range but then there are a whole lot of zeros following the top bit. So when we get a random if the top bit is set (50% chance) then if any of the other top bits are set (in the positions where there are all zeros below the top bit) then that random number will be too large and it will need to retry. To get around this it sees if the next 2 top bits are not set - and in this case it does magic to grab an extra rand bit and then try modulus on the bigger range.

Reply:
If any code requires the current behaviour, that code should probably get moved instead and get some comment of why it's needed.

Reply:
NOTE: this actually happens for some curves private key gen since the order has a tops bit set followed by lots of zeros.

Reply:
> A side effect of the new algorithm is that the output value is 1 + rand()... Which breaks some tests which feed the entropy in thru a buffer (which now needs to be subtracted by 1)

That is a very undesirable side-effect: that is the reason why in `ecdsatest.c` for the KAT we do 

```c
|| (fbytes_counter == 0 && !TEST_true(BN_sub_word(tmp, 1)))
```

which is confusing already now that these changes are presented together, and will simply be baffling when trying to understand the tests on their own in the future.

I would suggest to do, like done in most places already:

```c
/*-
 * INPUT
 *     range := our upper bound n
 *
 * OUTPUT:
 *     k: random log_2(range) bits long BN in [1, n-1], i.e. (0, n)
 */
do {
    if (!BN_priv_range(k, range)) {
        goto err;
    }

    /* BN_priv_range() provides k in [0, n) */
} while (BN_is_zero(k)); /* ensure k is not 0 */

/* here: 0 < k < n */
```

Reply:
That conforms to the old standard (kind of).  The issue is that they have actually changed the
process so I am not sure this would be classed as FIPS compliant (See SP800-56A R3 page 32 for (ECC) and Page 29 for (DH/DSA). [Especially with those range optimizations].
I could just ignore the changes and hope noone notices, but is that the best option? :)

Reply:
> That conforms to the old standard (kind of). The issue is that they have actually changed the
> process so I am not sure this would be classed as FIPS compliant (See SP800-56A R3 page 32 for (ECC) and Page 29 for (DH/DSA). [Especially with those range optimizations].
> I could just ignore the changes and hope noone notices, but is that the best option? :)

I would argue that it would be preferable to do the following:
- (in a separate PR) revise `BN_{,priv_}rand_range()` (independently of FIPS mode) to conform to `SP800-90A A.5.3` (i.e. replace the current implementation---that admittedly looks quite messy--- by instead requesting `log_2(range)+64` bits from the RBG and then doing `k := k mod(range)`
- for keygen here conform to `SP800-56A R3 5.6.1.2.1` (_Key Pair Generation Using Extra Random Bits_) rather than `5.6.1.2.2` (_Key Pair Generation by Testing Candidates_)---the one you are following now; once `BN_priv_rand_range()` is fixed as per the previous point, I claim doing `do { BN_priv_rand_range(k, range); } while (BN_is_zero(k));` is functionally equivalent to step 6 of `SP800-56A R3 5.6.1.2.1` to ensure the generated `k` is in `(0, n)` _exclusive_.

Reply:
SP800-56A seems to have 2 algorithms, one that takes 64 extra bits, one that takes the correct amount of bits. At first sight, we seem to do something very similar, didn't properly check, but we only take 1 extra bit, and decide between the 2 based on some condition. I assume that the we do is just some speed optimization, so that we don't need to generate many random numbers. We might just need to take the 64 extra bit instead of just 1 bit, and then do a real mod function to be compatible.

I'm not sure it's worth to have both algorithms in our code, and we could also just go for just one of them.

Reply:
We could also add some function that has both an lower and upper value, which really is a wrapper around the current function that first calls it with range=upper-lower, and adds lower to the result.

Reply:
My issue with adding/subtracting is that it will have unexpected consequences on any known answer test as we will lose the ability to control the scalar by feeding predetermined randomness into the RNG without including stuff in the tests that rely on internal implementation quirks

Reply:
It's my understanding that the 1 is added anyway, it would just be moved to a new function.

It's always somehow going to depend on implementation details.

Reply:
> It's my understanding that the 1 is added anyway, it would just be moved to a new function.

If you do the extra bits version and then modular reduction, you have 2 choices:
- do `k := longer_bn mod(range-1)` followed by `k := k+1`
- do `k := longer_bn mod(range)`, repeat if `k == 0` (_which happens with probability of 1/n_)

I would prefer the second one, in which no extra subtractions/additions are necessary.

Reply:
And my point is that either the k += 1 or the repeat if k == 0 is currently in some other function, not the one that returns you the number in a range.

If you want to generalize something, either it's a function that returns something in 1-range, or lower-upper. And if you go for the one with lower and upper, the addition looks like the most obvious way, but you could also just repeat until you have a number in the range.

Reply:
@paulidale and I have both looked at little closer at the current implementation of the rand range function - it does not appear to be biased. Dont know that it is worth trying to put in a new API just to do this fairly simple thing. 

I think the following is sufficient (i.e the old way)
    /*
     * Although this is slightly different from the standard it is effectively
     * equivalent as it gives an unbiased result ranging from 1..n-1. It is also
     * faster as the standard needs to retry more often. Also doing
     * 1 + rand[0..n-2) would effect the way that tests feed dummy entropy into
     * rand so the simpler backward compatible method has been chosen here.
     */
    do
    {
        if (!BN_priv_rand_range(priv_key, order))
            goto err;
    } while (BN_is_zero(priv_key));

Reply:
> And my point is that either the k += 1 or the repeat if k == 0 is currently in some other function, not the one that returns you the number in a range.

Ah, I missed your point then!
I was writing on the assumption of keeping the current `BN_rand_range()` API of returning something in `[0, range-1]` _inclusive_ and then having the caller dealing with the exclusion of the lower bound.

> If you want to generalize something, either it's a function that returns something in 1-range, or lower-upper. And if you go for the one with lower and upper, the addition looks like the most obvious way, but you could also just repeat until you have a number in the range.

For a new generalized function
```c
int BN_rand_range_ex(BIGNUM *out, const BIGNUM *min_inclusive, const BIGNUM *max_inclusive);
int BN_rand_priv_range_ex(BIGNUM *out, const BIGNUM *min_inclusive, const BIGNUM *max_inclusive);
```
as you say addition seems the obvious way, and could be ok for the KATs as well, as then the offsetting wouldn't be ad-hoc in certain modes or for very specific steps (e.g., in the current code of this PR we subtract 1 only when feeding the keygen and not for `sign_setup` that should generate a nonce using exactly the same logic of the keygen operation).

But again, this is one more reason why this PR should be split in more parts, even if the actual code changes seems small:
- we shouldn't have an ad-hoc `bn_rand_priv_range()` used only by the ECC keygen and that cannot be accessed externally by ENGINEs/providers or external applications;
- the proposed `bn_rand_priv_range()` (or the alternatives discussed in this thread) is not limited to _FIPS ECC related changes_
- the proposed changes to the ECC keygen method are not limited to FIPS mode
- changes in the ECC keygen should be mirrored in the `ecdsa_sign_setup()` as they generally have the same requirements
- changes to the `EC_POINT` comparison logic shouldn't be limited to FIPS mode and should stay unified
- some of the considerations on point validation should probably apply to non-FIPS mode too, while some should stay FIPS mode specific

Reply:
> SP800-56A seems to have 2 algorithms, one that takes 64 extra bits, one that takes the correct amount of bits.

I'll note the thread at https://mailarchive.ietf.org/arch/msg/cfrg/6k9HMqfmdiKDdFFUHG3IJOGaF8g that is perhaps timely, w.r.t. that 64 number of extra bits